### PR TITLE
Object store support for multiple stores and pool parameters

### LINKED ISSUE #310: rook command should support multiple named object stores
The rook command only support one object store.

It should be extended to support multiple named object stores and include the ability to "ls" an object store as well as query available properties.

### PR DESCRIPTION
**Pull request by travisn**
The object store now supports:
- Multiple instances of the object store, implemented by separate ceph realms
- Pool parameters for replication and erasure coding
- SSL certificate on the rgw instances

Fixes #310 and #334 and lays a foundation for #698

### CODE REVIEW

File: cmd/rookctl/object/create.go [hunk 1]
```diff
@@ -17,43 +17,91 @@ package object
...
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "failed to read certificate", err)
+			os.Exit(1)
+		}
+		store.Certificate = string(cert)
+	}
+
+	dataPool, err := pool.ConfigToModel(dataConfig)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
```
Comment (jbw976) [inline cmd/rookctl/object/create.go]: include message in the output about "failed to read data pool config"?  same below for metadata pool config

File: cmd/rookctl/object/create.go [hunk 2]
```diff
@@ -17,43 +17,91 @@ package object
...
+	metadataPool, err := pool.ConfigToModel(metadataConfig)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+	store.MetadataConfig = *metadataPool
 
-	c := rook.NewRookNetworkRestClient()
-	out, err := createObjectStore(c)
+	c := rook.NewRookNetworkRestClientWithTimeout(initObjectStoreTimeout * time.Second)
```
Comment (jbw976) [inline cmd/rookctl/object/create.go]: does the create happen synchronously now? previously, it was an async operation so no extended timeout would be needed here for Create.  Later operations, such as connection/user no longer have the extended timeout they used to, so I'm wondering how the flow works here now.

File: pkg/ceph/client/pool.go [hunk 3]
```diff
@@ -113,7 +113,7 @@ func CreatePool(context *clusterd.Context, clusterName string, newPool CephStora
 
 func CreatePoolForApp(context *clusterd.Context, clusterName string, newPool CephStoragePoolDetails, appName string) error {
 	args := []string{"osd", "pool", "create", newPool.Name, strconv.Itoa(newPool.Number)}
-	// not implemented: fix the pool create for the different profiles
```
Comment (jbw976) [inline pkg/ceph/client/pool.go]: how did this "not implemented" get fixed?

File: pkg/ceph/rgw/daemon.go [hunk 4]
```diff
@@ -65,6 +67,13 @@ func generateConfigFiles(context *clusterd.Context, config *Config) error {
 		logger.Warningf("failed to create data directory %s: %+v", dataDir, err)
 	}
 
+	sslSuffix := ""
+	if config.Certificate != "" {
+		// the suffix is intended to be appended to the end of the rgw_frontends arg, immediately after the port.
+		// with ssl enabled, the port number must end with the letter s.
+		sslSuffix = fmt.Sprintf("s ssl_certificate=%s", config.Certificate)
```
Comment (jbw976) [inline pkg/ceph/rgw/daemon.go]: i'm not sure if i traced this correctly, but is this `config.Certificate` field the binary data that represents the cert? If so, do we need to do any escaping or sanitization?

File: pkg/operator/rgw/rgw.go [hunk 5]
```diff
@@ -89,8 +110,141 @@ func (c *Cluster) Start() error {
...
+
+func (c *Cluster) createRealm(serviceIP string) error {
+	realmArg := fmt.Sprintf("--rgw-realm=%s", c.Config.Name)
+	zonegroupArg := fmt.Sprintf("--rgw-zonegroup=%s", c.Config.Name)
+	zoneArg := fmt.Sprintf("--rgw-zone=%s", c.Config.Name)
+	endpointArg := fmt.Sprintf("--endpoints=%s:%d", serviceIP, c.Config.Port)
+	updatePeriod := false
+
+	// create the realm if it doesn't exist yet
+	output, err := c.runRGWCommand("realm", "get", realmArg)
```
Comment (jbw976) [inline pkg/operator/rgw/rgw.go]: The `runRGWCommand` function doesn't have an output file to parse only the output from, it parses both stdout and stderr (combined output), right?  Are we at risk of getting malformed output when debugging logging is on, similar to #942?

File: pkg/operator/rgw/rgw.go [hunk 6]
```diff
@@ -89,8 +110,141 @@ func (c *Cluster) Start() error {
...
+			return fmt.Errorf("failed to create rgw zonegroup for %s. %+v", c.Config.Name, err)
+		}
+	}
+	zoneID, err := decodeID(output)
+	if err != nil {
+		return fmt.Errorf("failed to parse zone id. %+v", err)
+	}
+
+	if updatePeriod {
+		_, err = c.runRGWCommand("period", "update", "--commit")
```
Comment (jbw976) [inline pkg/operator/rgw/rgw.go]: what does update period do?  we are doing it conditionally, perhaps a comment on why and what it does could help

File: pkg/operator/rgw/rgw.go [hunk 7]
```diff
@@ -89,8 +110,141 @@ func (c *Cluster) Start() error {
...
+		updatePeriod = true
+		output, err = c.runRGWCommand("zonegroup", "create", "--master", zonegroupArg, realmArg, endpointArg)
+		if err != nil {
+			return fmt.Errorf("failed to create rgw zonegroup for %s. %+v", c.Config.Name, err)
+		}
+	}
+
+	zoneGroupID, err := decodeID(output)
+	if err != nil {
+		return fmt.Errorf("failed to parse realm id. %+v", err)
```
Comment (jbw976) [inline pkg/operator/rgw/rgw.go]: copy/paste, this error message is for zone group ID, right?

File: pkg/operator/rgw/rgw.go [hunk 8]
```diff
@@ -89,8 +110,141 @@ func (c *Cluster) Start() error {
...
+		if err != nil {
+			return fmt.Errorf("failed to update period. %+v", err)
+		}
+	}
+
+	logger.Infof("RGW: realm=%s, zonegroup=%s, zone=%s", realmID, zoneGroupID, zoneID)
+	return nil
+}
+
+func (c *Cluster) createPools() error {
```
Comment (jbw976) [inline pkg/operator/rgw/rgw.go]: is this idempotent? or does it only run once on CRD create event?

File: pkg/operator/rgw/rgw.go [hunk 9]
```diff
@@ -89,8 +110,141 @@ func (c *Cluster) Start() error {
...
+		if err := ceph.CreateErasureCodeProfile(c.context, c.Namespace, poolConfig.ErasureCodedConfig, cephConfig.ErasureCodeProfile); err != nil {
+			return fmt.Errorf("failed to create erasure code profile for object store %s: %+v", c.Config.Name, err)
+		}
+	}
+
+	for _, pool := range pools {
+		// create the pool if it doesn't exist yet
+		name := pool
+		if !strings.HasPrefix(pool, ".") {
+			name = fmt.Sprintf("%s.%s", c.Config.Name, pool)
```
Comment (jbw976) [inline pkg/operator/rgw/rgw.go]: it looks like only pools that start with `.` will get the object store name embedded into the pool name.  does that mean that the other pools are shared across object store instances?

File: cmd/rook/rgw.go [hunk 10]
```diff
@@ -32,14 +32,18 @@ var rgwCmd = &cobra.Command{
...
 	rgwHost    string
+	rgwCert    string
 	rgwPort    int
 )
 
 func init() {
+	rgwCmd.Flags().StringVar(&rgwName, "rgw-name", "", "name of the object store")
 	rgwCmd.Flags().StringVar(&rgwKeyring, "rgw-keyring", "", "the rgw keyring")
 	rgwCmd.Flags().StringVar(&rgwHost, "rgw-host", "", "dns host name")
+	rgwCmd.Flags().StringVar(&rgwCert, "rgw-cert", "", "path to the ssl certificate in pem format")
```
[the remark this answers is not in the document]
Comment (travisn) [reply cmd/rook/rgw.go]: `rookctl` is not aware of kubernetes, it only communicates with the rook api service and is for standalone scenarios. So `rookctl` reads the cert file, sends it to the REST API, then the api service stores it in a k8s secret, then creates the object store and rgw service to pick up on that secret. But users who create an object CRD directly will reference the secret, and the file will never be needed.

[the remark this answers is not in the document]
Comment (travisn) [reply cmd/rook/rgw.go]: and i'll look at the `csr` resource to see if it will work for us, thanks

File: cmd/rookctl/object/create.go (hunk 1 above)
Comment (travisn) [reply cmd/rookctl/object/create.go]: the message comes back quite descriptive, but you're right it would help clarify for data or metadata pool with a message

File: cmd/rookctl/object/create.go (hunk 2 above)
Comment (travisn) [reply cmd/rookctl/object/create.go]: correct, create is synchronous now. All the pools are created before this call returns, so there is no delay later for the other calls.

File: pkg/ceph/client/pool.go (hunk 3 above)
Comment (travisn) [reply pkg/ceph/client/pool.go]: this seemed like an old comment since we are creating both replicated and ec pools now. What other profiles is this referring to?

File: pkg/ceph/rgw/daemon.go (hunk 4 above)
Comment (travisn) [reply pkg/ceph/rgw/daemon.go]: `config.Certificate` is actually a file path that has been loaded into the pod by mounting the secret as a file. I'll add a comment

File: pkg/operator/rgw/rgw.go (hunk 5 above)
Comment (travisn) [reply pkg/operator/rgw/rgw.go]: good question, i'll confirm

File: pkg/operator/rgw/rgw.go (hunk 7 above)
Comment (travisn) [reply pkg/operator/rgw/rgw.go]: good catch

File: pkg/operator/rgw/rgw.go (hunk 6 above)
Comment (travisn) [reply pkg/operator/rgw/rgw.go]: will add a comment. It's a way of updating an epoch so interested parties can be aware when a realm or zone is created

File: pkg/operator/rgw/rgw.go (hunk 8 above)
Comment (travisn) [reply pkg/operator/rgw/rgw.go]: it is idempotent. Later down the stack it will check if the pool exists before creating it.

File: pkg/operator/rgw/rgw.go (hunk 9 above)
Comment (travisn) [reply pkg/operator/rgw/rgw.go]: right, there is a single pool named `.rgw.root` that tracks across all object stores. I'll add a comment

File: pkg/ceph/client/pool.go (hunk 3 above)
Comment (jbw976) [reply pkg/ceph/client/pool.go]: yeah it looks like that comment came from the refactor to use external ceph tools.  as long as both EC and replicated pools are working, I think we're OK now.

File: pkg/operator/rgw/rgw.go (hunk 9 above)
Comment (jbw976) [reply pkg/operator/rgw/rgw.go]: ah oops, i missed the `!`, you're right that it's just the sole `.rgw.root` that is shared across object stores.  sounds good.

File: cmd/rook/rgw.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (travisn) [reply cmd/rook/rgw.go]: The format that ceph expects the pem (in a single file) and the format expected for the csr seems to be different. If I try to create a csr with the ceph-compatible pem, I get an error `PEM block type must be CERTIFICATE REQUEST`. It's not clear how to get the ceph-compatible pem to work as a csr.

File: pkg/ceph/rgw/admin.go [hunk 11]
```diff
@@ -49,7 +49,7 @@ func RunAdminCommand(c *Context, command, subcommand string, args ...string) (st
 	options = append(options, args...)
 
 	// start the rgw admin command
-	output, err := c.context.Executor.ExecuteCommandWithCombinedOutput(false, "", "radosgw-admin", options...)
+	output, err := c.context.Executor.ExecuteCommandWithOutput(false, "", "radosgw-admin", options...)
```
Comment (jbw976) [inline pkg/ceph/rgw/admin.go]: what impact does losing stderr have on other usages of `RunAdminCommand`, such as list buckets, delete bucket, create user?

do we still get a useful error message if those were to fail?

Comment (travisn) [reply pkg/ceph/rgw/admin.go]: i'm wondering if we should wait on this change until there is a fix upstream for the output file. For now there are two approaches. 
- Use the combined output. If debug logging is enabled it will fail. But this will only be an issue for the operator and api pods. But those pods probably aren't where the ceph debug logging would be enabled anyway.  We wait for a fix upstream to get an out-file option. 
- Ignore the stderr. On failure we could retry with stderr captured so we could have the log. Seems hackish to retry just for the stderr.

I'm leaning towards the first option of doing nothing since it's not a scenario to enable the debug logging in the operator/api pods.

File: cmd/rookctl/pool/create.go [hunk 12]
```diff
@@ -82,31 +110,36 @@ func createPoolsEntry(cmd *cobra.Command, args []string) error {
...
 	}
 
-	newPool := model.Pool{Name: poolName}
+	newPool := &model.Pool{}
 
-	if poolType == PoolTypeReplicated {
-		if dataChunks > 0 || codingChunks > 0 {
-			return "", fmt.Errorf("both data chunks and coding chunks must be zero for pool type '%s'", PoolTypeReplicated)
+	if config.PoolType == PoolTypeReplicated {
+		if config.DataChunks > 0 || config.CodingChunks > 0 {
```
Comment (kokhang) [inline cmd/rookctl/pool/create.go]: nit: Use != 0 otherwise negative numbers will make this condition to work

File: cmd/rookctl/pool/create_test.go [hunk 13]
```diff
@@ -121,7 +125,8 @@ func TestCreatePoolFailure(t *testing.T) {
 		},
 	}
 
-	out, err := createPool("pool1", PoolTypeReplicated, 0, 0, 0, c)
+	pool := model.Pool{Name: "pool1", Type: model.Replicated}
+	out, err := createPool(pool, c)
```
Comment (kokhang) [inline cmd/rookctl/pool/create_test.go]: perhaps add a test with negative datachunk or coding chunk values?

File: pkg/api/object.go [hunk 14]
```diff
@@ -25,12 +25,41 @@ import (
...
+
+	// only the default store is supported through the rest api
+	if objectStore.Name == "" {
+		objectStore.Name = defaultObjectStoreName
+	}
+	if objectStore.Port == 0 {
+		objectStore.Port = model.RGWPort
+	}
+	if objectStore.RGWReplicas == 0 {
+		objectStore.RGWReplicas = 1
```
Comment (kokhang) [inline pkg/api/object.go]: is 1 the default replica size if not specified? If so, it should be defined as a constant.

File: pkg/api/object.go [hunk 15]
```diff
@@ -44,7 +73,7 @@ func (h *Handler) CreateObjectStore(w http.ResponseWriter, r *http.Request) {
 // DELETE
 // /objectstore
 func (h *Handler) RemoveObjectStore(w http.ResponseWriter, r *http.Request) {
-	if err := h.config.ClusterHandler.RemoveObjectStore(); err != nil {
+	if err := h.config.ClusterHandler.RemoveObjectStore(defaultObjectStoreName); err != nil {
```
Comment (kokhang) [inline pkg/api/object.go]: what if the object store was created with an user-specified name? This will fail to delete.

File: pkg/api/object.go [hunk 16]
```diff
@@ -59,7 +88,7 @@ func (h *Handler) RemoveObjectStore(w http.ResponseWriter, r *http.Request) {
 // /objectstore/connectioninfo
 func (h *Handler) GetObjectStoreConnectionInfo(w http.ResponseWriter, r *http.Request) {
```
Comment (kokhang) [inline pkg/api/object.go]: same here with user-specified name

File: pkg/ceph/rgw/admin.go [hunk 17]
```diff
@@ -23,21 +23,33 @@ import (
...
 
-func RunAdminCommand(context *clusterd.Context, getClusterInfo func() (*mon.ClusterInfo, error), command, subcommand string, args ...string) (string, error) {
-	cluster, err := getClusterInfo()
+type Context struct {
+	context        *clusterd.Context
+	objectStore    string
+	getClusterInfo func() (*mon.ClusterInfo, error)
+}
+
+func NewContext(context *clusterd.Context, objectStore string, getClusterInfo func() (*mon.ClusterInfo, error)) *Context {
```
Comment (kokhang) [inline pkg/ceph/rgw/admin.go]: shouldnt this be added to the clusterd.context instead? That way we have one context object with all data for the cluster.

File: cmd/rookctl/pool/create.go (hunk 12 above)
Comment (travisn) [reply cmd/rookctl/pool/create.go]: negative numbers could be ignored just like 0

File: pkg/api/object.go (hunk 15 above)
Comment (travisn) [reply pkg/api/object.go]: it wasn't plumbed yet for the rest api to support multiple. I'm adding that in the next PR

File: pkg/ceph/rgw/admin.go (hunk 17 above)
Comment (travisn) [reply pkg/ceph/rgw/admin.go]: Everything shouldn't go in the core context. The object-specific context is only needed here in rgw

File: pkg/api/object.go (hunk 14 above)
Comment (travisn) [reply pkg/api/object.go]: sure

File: cmd/rookctl/pool/create_test.go (hunk 13 above)
Comment (travisn) [reply cmd/rookctl/pool/create_test.go]: negative values will be ignored just like 0, so we're ok
